tests/test_edgeql_syntax.py

```diff
@@ -2609,6 +2609,14 @@ def test_edgeql_syntax_updatefor_01(self):
         UNION (UPDATE Foo FILTER (Foo.id = x.0) SET {bar := x.1});
         """
 
+    @tb.must_fail(errors.EdgeQLSyntaxError,
+                  r"Unexpected '}'",
+                  line=2, col=27)
+    def test_edgeql_syntax_empty_for_01(self):
+        """
+        SELECT (FOR x in {} UNION ());
+        """
```

Comment:
This bans a simple case, what about

```
WITH set := <schema::Object>{}
SELECT (FOR x in set UNION ());
```

?

Wouldn't it be better to fix the underlying problem instead of banning this at the grammar level?

cc @elprans

Reply:
@elplans suggested the syntactic approach because your snippet isn't valid syntax: the set in a FOR needs to be a set-literal.

That said, `WITH s := {} SELECT (FOR x in {s} UNION ());` triggers the same crash, so I'm going to abandon the syntactic approach.

Reply:
It was my suggestion to ban this at the syntax level because `FOR` requires a set literal syntactically (otherwise we have ambiguous grammar).  The underlying problem is that the variable you iterate over in `FOR` must have a concrete type and an empty set literal `{}` is of type `anytype`.  Since you can't put a cast _before_ the empty set literal in `FOR` it makes sense to just make it a syntax error.

That said, we still need to add a check that the `FOR` variable is not of a generic type (your example shows how one can still trigger the bug).

Reply:
@msullivan 
> the set in a FOR needs to be a set-literal.

Right!

@elprans 
> It was my suggestion to ban this at the syntax level because FOR requires a set literal syntactically (otherwise we have ambiguous grammar). 

Alright. I guess we can resurrect this PR with the approach I suggested in https://github.com/edgedb/edgedb/pull/1602#pullrequestreview-456944615

But the `WITH s := {} SELECT (FOR x in {s} UNION ());` case should also be fixed, ideally, so maybe we try to fix both things in one place.